framework/src/iotbus/iotbus_uart.c

```diff
@@ -395,7 +418,191 @@ int iotbus_uart_write(iotbus_uart_context_h hnd, const char *buf, unsigned int l
 	return ret;
 }
 
-#ifdef CONFIG_IOTDEV
+static int _uart_sig[IOTBUS_UART_INTR_MAX] = {
+	SIG_IOTBUS_UART_TX_EMPTY,
+	SIG_IOTBUS_UART_TX_RDY,
+	SIG_IOTBUS_UART_RX_AVAIL,
+	SIG_IOTBUS_UART_RECEIVED
+};
+
+static void signal_handler(int sig, siginfo_t *info, void *ctx)
+{
+	void *ptr;
+	iotbus_uart_context_h dev;
+	struct _iotbus_uart_s *handle;
+	int type;
+
+ #ifdef CONFIG_CAN_PASS_STRUCTS
+	ptr = info->si_value.sival_ptr;
+#else
+	ptr = info->si_value;
+#endif
+	if (ptr == NULL) {
+		ibdbg("Handler is not assigned.\n");
+		return;
+	}
+	dev = (iotbus_uart_context_h)ptr;
+	handle = (struct _iotbus_uart_s *)dev->handle;
+
+	switch (sig) {
+	case SIG_IOTBUS_UART_TX_EMPTY:
+		type = IOTBUS_UART_TX_EMPTY;
+		break;
+	case SIG_IOTBUS_UART_TX_RDY:
+		type = IOTBUS_UART_TX_RDY;
+		break;
+	case SIG_IOTBUS_UART_RX_AVAIL:
+		type = IOTBUS_UART_RX_AVAIL;
+		break;
+	case SIG_IOTBUS_UART_RECEIVED:
+		type = IOTBUS_UART_RECEIVED;
+		break;
+	default:
+		ibdbg("signo is wrong.\n");
+		return;
+	}
+	
+	// Call callback function.
+	if (handle->cb[type] != NULL) {
+		((iotbus_uart_cb)handle->cb[type])(dev);
+	}
+}
+
+static pthread_addr_t uart_intr_thread(pthread_addr_t arg)
+{
+	iotbus_uart_intr_e int_type = (iotbus_uart_intr_e)(unsigned long)arg;
+	if (int_type < 0 || int_type >= IOTBUS_UART_INTR_MAX) {
+		ibdbg("Error : invalid int type\n");
+		pthread_exit(NULL);
+		return NULL;
+	}
+	ibdbg("Thread for [%d]\n", int_type);
+
+	struct sigaction sigact;
+	sigact.sa_sigaction = signal_handler;
+	(void)sigfillset(&sigact.sa_mask);
+	(void)sigdelset(&sigact.sa_mask, _uart_sig[int_type]);
+	sigact.sa_flags = SA_SIGINFO;
+
+	sigaction(_uart_sig[int_type], &sigact, NULL);
+
+	// Not to finish this thread.
+	while (1) { sleep(1); }
```

Comment:
As I remember I gave to you that we have to reduce thread as possible as we can.
You made thread for all peripheral, and they will awake for every 1 sec because of this code.(I think this is very bad code)
I think you had better consider below.
1. Make just one thread as a common interrupt handling
2. Provide api & structure for priority of interrupt (Need to be processed immediately or can wait...?) also queue.
3. Whatever you use poll or signal put it to queue.
You can refer to message queue in freertos.